langchain/tools/pipeline/tool.py

```diff
@@ -0,0 +1,132 @@
+    output_expression: Any,
+    out: Any,
+    extra: Optional[Union[str, Dict[str, Any]]] = None,
+    final: bool = False,
+) -> Any:
+    if output_expression is None:
+        return out
+    if extra:
+        extra = copy(extra)
+
+    if final:
+        rval = {}
+    else:
+        rval = copy(extra) if isinstance(extra, dict) else {}
+
+    if isinstance(output_expression, str):
+        rval[output_expression] = out
+    elif isinstance(output_expression, (tuple, list)):
+        for k in output_expression:
+            rval[k] = out[k]
+    else:
+        for k, expression in output_expression.items():
+            rval[k] = jmespath.search(expression, out)
+
+    if final and len(rval) == 1:
+        return list(rval.values())[0]
+
+    return rval
+
+
+class PipelineStep(BaseModel):
+    """Base class for a pipeline step."""
+
+    func: Callable
+    input_expression: Optional[Union[str, Dict[str, str]]] = None
+    output_expression: Optional[Union[str, List[str], Dict[str, str]]] = None
+
+    def __call__(
+        self, input_: Union[str, Dict[str, Any]]
+    ) -> Union[str, Dict[str, Any]]:
+        """Execute the step function.
+
+        Args:
+            input_ (Union[str, Dict[str, Any]]): The input to the step function.
+
+        Raises:
+            ValueError: Raised when there is a mismatch between the input and input_expression.
+
+        Returns:
+            Union[str, Dict[str, Any]]: The extracted output based on the output_expression.
+        """
+        if self.input_expression is not None:
+            if isinstance(input_, str):
+                try:
+                    input_ = json.loads(input_)
+                except json.JSONDecodeError:
+                    pass
+
+            if not isinstance(input_, dict):
+                raise ValueError(
+                    "Input value should be a dict if input_expression is available"
+                )
+
+            args, kw = [], {}
+            if isinstance(self.input_expression, dict):
+                for key, expression in self.input_expression.items():
+                    value = jmespath.search(expression, input_)
+                    kw[key] = value
+            else:
+                kw[self.input_expression] = input_[self.input_expression]
+        if not self.input_expression:
+            args, kw = [input_], {}
+
+        out = self.func(*args, **kw)
+        return _return_based_on_expression(self.output_expression, out, extra=input_)
+
+
+class PipelineTool(BaseTool):
+    """The Pipeline Tool is designed to facilitate the creation of pipelines.
+
+    It allows for the utilization of existing agents, chains, and tools
+    within each step of the pipeline.
+    """
+
+    name = "pipeline_tool"
+    description = """
+    Can be used to construct a context task handler with existing tools and chains.
+    """
+
+    steps: List[PipelineStep]
+    output_expression: Optional[Union[str, List[str], Dict[str, str]]] = None
```

Comment:
Could you clarify what is the purpose of output_expression? And could it be implemented as just another step in the pipeline? If so I'd be inclined to remove it